File: cli/pb_cli/core/dead_code.py
"""Dead code reachability analysis — pure BFS over a call graph.

Pure module — no I/O, no DuckDB.

Public API:
    compute_dead_procedures(procedures, calls, resolved, inherits, dw_objects)
        -> list[DeadProcedure]
"""

from __future__ import annotations

from collections import defaultdict, deque
from typing import NamedTuple


class DeadProcedure(NamedTuple):
    object: str
    name: str
    proc_type: str
    cyclomatic: int | None
    confidence: str  # 'high' | 'medium' | 'low'
    caller_count_naive: int
    caller_count_scoped: int


def compute_dead_procedures(
    procedures: list[tuple[str, str, str, int | None]],
    calls: list[tuple[str, str, str]],
    resolved: list[tuple[str, str, str, str]],
    inherits: list[tuple[str, str]],
    dw_objects: set[str],
) -> list[DeadProcedure]:
    """Compute unreachable procedures via BFS from entry points.

    Args:
        procedures: (object, name, proc_type, cyclomatic)
        calls: (object, from_proc, to_name)
        resolved: (object, from_proc, target_object, target_proc)
        inherits: (from_object, to_object)
        dw_objects: set of DataWindow object names

    Returns:
        Sorted list of dead procedures with confidence classification.
    """
    # Index procedures by (object, name)
    proc_index: dict[tuple[str, str], tuple[str, int | None]] = {}
    for obj, name, ptype, cc in procedures:
        proc_index[(obj, name)] = (ptype, cc)

    # Build same-object edges: caller -> callee when callee lives in same object.
    # Case-insensitive: lower(to_name) matches lower(procedure name).
    same_obj_by_proc: dict[tuple[str, str], list[tuple[str, str]]] = defaultdict(list)
    # Index procedure names by object for efficient lookup
    procs_by_obj: dict[str, dict[str, str]] = defaultdict(dict)
    for (p_obj, p_name) in proc_index:
        procs_by_obj[p_obj][p_name.lower()] = p_name

    for obj, from_proc, to_name in calls:
        name_map = procs_by_obj.get(obj)
        if name_map:
            matched = name_map.get(to_name.lower())
            if matched:
                same_obj_by_proc[(obj, from_proc)].append((obj, matched))

    # Build cross-object edges from resolved_calls.
    cross_obj: dict[tuple[str, str], list[tuple[str, str]]] = defaultdict(list)
    for obj, from_proc, tgt_obj, tgt_proc in resolved:
        cross_obj[(obj, from_proc)].append((tgt_obj, tgt_proc))

    # Build override edges: if parent.m is reachable, child.m (override) is too.
    children_of: dict[str, list[str]] = defaultdict(list)
    for child, parent in inherits:
        children_of[parent].append(child)

    methods_by_obj: dict[str, set[str]] = defaultdict(set)
    for (obj, name) in proc_index:
        methods_by_obj[obj].add(name)

    override_edges: dict[tuple[str, str], list[tuple[str, str]]] = defaultdict(list)
    for parent_obj, methods in methods_by_obj.items():
        for child_obj in children_of.get(parent_obj, []):
            for method in methods:
                if (child_obj, method) in proc_index:
                    override_edges[(parent_obj, method)].append((child_obj, method))

    # Combine all edges into adjacency list.
    edges: dict[tuple[str, str], list[tuple[str, str]]] = defaultdict(list)
    for key, targets in same_obj_by_proc.items():
        edges[key].extend(targets)
    for key, targets in cross_obj.items():
        edges[key].extend(targets)
    for key, targets in override_edges.items():
        edges[key].extend(targets)

    # Seed 1: event and on handlers are always reachable.
    seeds: set[tuple[str, str]] = set()
    for obj, name, ptype, _cc in procedures:
        if ptype in ("event", "on"):
            seeds.add((obj, name))

    # Seed 2: procedures in DW objects that have calls are reachable.
    for obj, from_proc, _to_name in calls:
        if obj in dw_objects:
            seeds.add((obj, from_proc))

    # BFS from seeds.
    reachable: set[tuple[str, str]] = set()
    queue: deque[tuple[str, str]] = deque()
    for seed in seeds:
        if seed not in reachable:
            reachable.add(seed)
            queue.append(seed)

    while queue:
        current = queue.popleft()
        for neighbor in edges.get(current, ()):
            if neighbor not in reachable:
                reachable.add(neighbor)
                queue.append(neighbor)

    # Compute caller counts for dead procedures.
    naive_map: dict[str, set[tuple[str, str]]] = defaultdict(set)
    for obj, from_proc, to_name in calls:
        naive_map[to_name.lower()].add((obj, from_proc))

    scoped_map: dict[tuple[str, str], int] = defaultdict(int)
    for _obj, _from_proc, tgt_obj, tgt_proc in resolved:
        scoped_map[(tgt_obj, tgt_proc)] += 1

    # Collect dead procedures.
    dead: list[DeadProcedure] = []
    for (obj, name), (ptype, cc) in proc_index.items():
        if (obj, name) in reachable:
            continue
        naive = len(naive_map.get(name.lower(), set()))
        scoped = scoped_map.get((obj, name), 0)
        if naive == 0:
            confidence = "high"
        elif scoped == 0:
            confidence = "medium"
        else:
            confidence = "low"
        dead.append(DeadProcedure(obj, name, ptype, cc, confidence, naive, scoped))

    dead.sort(key=lambda d: (d.object, d.name))
    return dead

File: parser/src/PB/Pipeline/DeadCode.hs
module PB.Pipeline.DeadCode
  ( DeadProcedure (..)
  , ProcInfo (..)
  , computeDeadProcedures
  , cyclomaticComplexity
  ) where

import PB.Prelude
import PB.Pipeline.CfgBuild  (Cfg (..))
import Data.List              (sortOn)
import Data.Map.Strict qualified as Map
import Data.Set qualified as Set
import Data.Text qualified as T

-- | A procedure in the workspace.
data ProcInfo = ProcInfo
  { piObject   :: Text
  , piName     :: Text
  , piProcType :: Text   -- "function" | "subroutine" | "event" | "on"
  , piCyclomatic :: Maybe Int
  } deriving (Eq, Show)

-- | A dead (unreachable) procedure with confidence classification.
data DeadProcedure = DeadProcedure
  { dpObject          :: Text
  , dpName            :: Text
  , dpProcType        :: Text
  , dpCyclomatic      :: Maybe Int
  , dpConfidence      :: Text   -- "high" | "medium" | "low"
  , dpCallerCountNaive  :: Int
  , dpCallerCountScoped :: Int
  } deriving (Eq, Show)

-- | Compute cyclomatic complexity from a CFG: E - N + 2*P.
-- For a single connected procedure body, P = 1, so complexity = E - N + 2.
cyclomaticComplexity :: Cfg -> Int
cyclomaticComplexity cfg =
  let n = length (cfgBlocks cfg)
      e = length (cfgEdges cfg)
  in  if n == 0 then 1 else e - n + 2

-- | Compute unreachable procedures via BFS from entry points.
--
-- Inputs:
--   procedures — all procedures in the workspace
--   calls      — raw call sites (object, from_proc, to_name)
--   resolved   — resolved calls (object, from_proc, target_object, target_proc)
--   inherits   — inheritance edges (child, parent)
--   dwObjects  — set of DataWindow object names
computeDeadProcedures
  :: [ProcInfo]              -- ^ all procedures
  -> [(Text, Text, Text)]    -- ^ raw calls: (object, from_proc, to_name)
  -> [(Text, Text, Text, Text)] -- ^ resolved calls: (object, from_proc, target_object, target_proc)
  -> [(Text, Text)]          -- ^ inherits: (child, parent)
  -> Set.Set Text            -- ^ DataWindow object names
  -> [DeadProcedure]
computeDeadProcedures procedures calls resolved inherits dwObjects =
  let -- Index procedures by (object, name)
      procIndex = Map.fromList
        [ ((piObject p, piName p), p) | p <- procedures ]

      -- Index procedure names by object (case-insensitive lookup)
      procsByObj :: Map.Map Text (Map.Map Text Text)
      procsByObj = Map.fromListWith Map.union
        [ (piObject p, Map.singleton (T.toLower (piName p)) (piName p))
        | p <- procedures
        ]

      -- Build same-object edges: caller -> callee when callee lives in same object
      sameObjEdges = Map.fromListWith (++)
        [ ((obj, fromProc), [(obj, matched)])
        | (obj, fromProc, toName) <- calls
        , nameMap <- maybeToList (Map.lookup obj procsByObj)
        , matched <- maybeToList (Map.lookup (T.toLower toName) nameMap)
        ]

      -- Build cross-object edges from resolved calls
      crossObjEdges = Map.fromListWith (++)
        [ ((obj, fromProc), [(tgtObj, tgtProc)])
        | (obj, fromProc, tgtObj, tgtProc) <- resolved
        ]

      -- Build override edges: if parent.m is reachable, child.m is too
      childrenOf = Map.fromListWith (++)
        [ (parent, [child]) | (child, parent) <- inherits ]

      methodsByObj = Map.fromListWith Set.union
        [ (piObject p, Set.singleton (piName p)) | p <- procedures ]

      overrideEdges = Map.fromListWith (++)
        [ ((parentObj, method), [(childObj, method)])
        | (parentObj, methods) <- Map.toList methodsByObj
        , childObj <- Map.findWithDefault [] parentObj childrenOf
        , method <- Set.toList methods
        , (childObj, method) `Map.member` procIndex
        ]

      -- Combine all edges
      allEdges = Map.unionWith (++) (Map.unionWith (++) sameObjEdges crossObjEdges) overrideEdges

      -- Seed 1: event and on handlers are always reachable
      seeds = Set.fromList
        [ (piObject p, piName p)
        | p <- procedures
        , piProcType p `elem` ["event", "on"]
        ]

      -- Seed 2: procedures in DW objects that have calls are reachable
      dwSeeds = Set.fromList
        [ (obj, fromProc)
        | (obj, fromProc, _) <- calls
        , obj `Set.member` dwObjects
        ]

      allSeeds = Set.union seeds dwSeeds

      -- BFS from seeds
      reachable = bfs allSeeds allEdges

      -- Compute caller counts for dead procedures
      naiveMap = Map.fromListWith Set.union
        [ (T.toLower toName, Set.singleton (obj, fromProc))
        | (obj, fromProc, toName) <- calls
        ]

      scopedMap = Map.fromListWith (+)
        [ ((tgtObj, tgtProc), 1 :: Int)
        | (_, _, tgtObj, tgtProc) <- resolved
        ]

      -- Collect dead procedures (deduplicate on (object, name) —
      -- overloaded functions may produce multiple ProcInfo entries).
      deadMap = Map.fromListWith (\a _b -> a)
        [ ((piObject p, piName p), DeadProcedure
              { dpObject = piObject p
              , dpName = piName p
              , dpProcType = piProcType p
              , dpCyclomatic = piCyclomatic p
              , dpConfidence =
                  let naive = Set.size (Map.findWithDefault Set.empty (T.toLower (piName p)) naiveMap)
                      scoped = Map.findWithDefault 0 (piObject p, piName p) scopedMap
                  in if naive == 0 then "high"
                     else if scoped == 0 then "medium"
                     else "low"
              , dpCallerCountNaive =
                  Set.size (Map.findWithDefault Set.empty (T.toLower (piName p)) naiveMap)
              , dpCallerCountScoped =
                  Map.findWithDefault 0 (piObject p, piName p) scopedMap
              })
        | p <- procedures
        , (piObject p, piName p) `Set.notMember` reachable
        ]
      dead = sortOn (\d -> (dpObject d, dpName d)) (Map.elems deadMap)
  in  sortOn (\d -> (dpObject d, dpName d)) dead

-- | BFS reachability from seed set through adjacency map.
bfs :: Set.Set (Text, Text) -> Map.Map (Text, Text) [(Text, Text)] -> Set.Set (Text, Text)
bfs seeds edges = go seeds (Set.toList seeds)
  where
    go visited [] = visited
    go visited (current:rest) =
      let neighbors = Map.findWithDefault [] current edges
          newNeighbors = filter (`Set.notMember` visited) neighbors
          visited' = foldl' (flip Set.insert) visited newNeighbors
      in  go visited' (rest ++ newNeighbors)
